d3a_interface/sim_results/kpi.py

```diff
@@ -145,11 +146,78 @@ def update_area_kpi(self, area_dict, core_stats):
+            area_dict: contain nested area info
+            core_stats: contain area"s raw/key statistics
+            gf_alp: grid_fee_along_the_path - cumulative grid fee from root to target area
+
+        """
+        self.populate_consumer_producer_sets(area_dict)
+
+        # fir_excl_gf_alp: feed-in tariff rate excluding grid fee along path
+        fir_excl_gf_alp = self.get_feed_in_tariff_rate_excluding_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        # mmr_incl_gf_alp: market maker rate include grid fee along path
+        mmr_incl_gf_alp = self.get_market_maker_rate_including_path_grid_fees(
+            core_stats.get(area_dict["uuid"], {}), gf_alp)
+        for trade in core_stats.get(area_dict["uuid"], {}).get("trades", []):
+            if trade["seller_origin_id"] in self.producer_ess_set:
+                self.fit_revenue += fir_excl_gf_alp * trade["energy"]
+                self.d3a_cost -= trade["price"]
+            if trade["buyer_origin_id"] in self.consumer_ess_set and \
+                    trade["seller_origin_id"] not in self.producer_ess_set:
+                self.utility_bill += mmr_incl_gf_alp * trade["energy"]
+                self.d3a_cost += trade["price"]
+        self.base_case_cost = self.utility_bill - self.fit_revenue
+        self.saving_absolute = self.base_case_cost - self.d3a_cost
+        self.saving_percentage = ((self.saving_absolute / self.base_case_cost) * 100
+                                  if self.base_case_cost else 0.)
+
+    def populate_consumer_producer_sets(self, area_dict):
+        for child in area_dict["children"]:
+            if is_producer_node_type(child):
+                self.producer_ess_set.add(child["uuid"])
+            elif is_load_node_type(child):
+                self.consumer_ess_set.add(child["uuid"])
+            elif is_prosumer_node_type(child):
+                self.producer_ess_set.add(child["uuid"])
+                self.consumer_ess_set.add(child["uuid"])
+
+    @staticmethod
+    def get_feed_in_tariff_rate_excluding_path_grid_fees(area_core_stat, path_grid_fee):
+        return area_core_stat.get("feed_in_tariff", 0.) - path_grid_fee
+
+    @staticmethod
+    def get_market_maker_rate_including_path_grid_fees(area_core_stat, path_grid_fee):
+        return area_core_stat.get("market_maker_rate", 0.) + path_grid_fee
+
+    def to_dict(self):
+        return {"base_case_cost": self.base_case_cost,
+                "d3a_cost": self.d3a_cost,
+                "saving_absolute": self.saving_absolute,
+                "saving_percentage": self.saving_percentage}
+
+
 class KPI(ResultsBaseClass):
     def __init__(self):
         self.performance_indices = dict()
         self.performance_indices_redis = dict()
         self.state = {}
+        self.savings_state = {}  # keeps initialized savings kpi object for individual area
+        self.area_uuid_to_cum_fee_path = {}  # keeps track of grid fee from root to target area
```

Comment:
Minor:
Sorry, i just saw this variable now. I found the name and the description a bid confusing. Hier a proposal for change (or you have a better idea).
```suggestion
        self.area_uuid_cum_grid_fee_mapping = {}  # mapping of area_uuid and the corresponding cumulative grid-free along the path from root-area to the area 
```